From 821a9035d61675e9436229f089c41183d852a16e Mon Sep 17 00:00:00 2001
From: fwaters <fwaters@anaconda.com>
Date: Mon, 23 Sep 2019 18:29:38 +0000
Subject: [PATCH 01/22] intel_mkl-version

---
 numpy/__init__.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/numpy/__init__.py b/numpy/__init__.py
index ba88c733f..3723daaf7 100644
--- a/numpy/__init__.py
+++ b/numpy/__init__.py
@@ -192,6 +192,8 @@ else:
     del PytestTester
 
 
+    __mkl_version__ = "__MKL_VERSION__"
+
     def _sanity_check():
         """
         Quick sanity checks for common bugs caused by environment.
-- 
2.20.1

