From bbdc4497959f82bf83da390a7009b3c945c7267c Mon Sep 17 00:00:00 2001
From: Sophia Castellarin <scastellarin@anaconda.com>
Date: Fri, 9 Aug 2019 00:02:39 -0500
Subject: [PATCH] remove specific compiler flags

---
 numpy/core/setup.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/numpy/core/setup.py b/numpy/core/setup.py
index c73e9c4a7..c20ff0f2e 100644
--- a/numpy/core/setup.py
+++ b/numpy/core/setup.py
@@ -931,6 +931,7 @@ def configuration(parent_package='',top_path=None):
             eca.append('/Qprec-div')
     else:
         eca = ['-fp-model', 'fast=2', '-fimf-precision=high', '-prec-sqrt']
+    eca = []
     config.add_library('loops',
                        sources=loops_src,
                        include_dirs=[],
-- 
2.20.1

